iree/compiler/Conversion/LinalgToSPIRV/FlattenMemRefSubspanPass.cpp

```diff
@@ -0,0 +1,329 @@
+
+    Value linearIndex =
+        linearizeIndices(storeOp.getMemRefType(), storeOp.getIndices(),
+                         storeOp.getLoc(), rewriter);
+    rewriter.replaceOpWithNewOp<memref::StoreOp>(storeOp, adaptor.value(),
+                                                 adaptor.memref(), linearIndex);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Folding Patterns
+//===----------------------------------------------------------------------===//
+
+/// Returns the number of bytes of the given `type`. Returns llvm::None if
+/// cannot deduce.
+///
+/// Note that this should be kept consistent with how the byte offset was
+/// calculated in the subspan ops!
+Optional<int64_t> getNumBytes(Type type) {
+  if (type.isIntOrFloat()) return (type.getIntOrFloatBitWidth() + 7) / 8;
+  if (auto vectorType = type.dyn_cast<VectorType>()) {
+    auto elementBytes = getNumBytes(vectorType.getElementType());
+    if (!elementBytes) return llvm::None;
+    return elementBytes.getValue() * vectorType.getNumElements();
+  }
+  return llvm::None;
+}
+
+/// Folds the byte offset on subspan ops into the consumer load/store ops.
+template <typename OpType>
+struct FoldSubspanOffsetIntoLoadStore final : public OpRewritePattern<OpType> {
+  using OpRewritePattern<OpType>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    auto memrefType = op.memref().getType().template cast<MemRefType>();
+    if (!isRankZeroOrOneMemRef(memrefType)) return failure();
+
+    auto subspanOp =
+        op.memref()
+            .template getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!subspanOp) return failure();
+
+    // If the subspan op has a zero byte offset then we are done.
+    if (matchPattern(subspanOp.byte_offset(), m_Zero())) return failure();
+    // byte length is unsupported for now.
+    if (subspanOp.byte_length()) return failure();
+
+    // Create a new subspan op with zero byte offset.
+    Value zero = rewriter.create<ConstantIndexOp>(op.memref().getLoc(), 0);
+    Value newSubspan = rewriter.create<IREE::HAL::InterfaceBindingSubspanOp>(
+        op.memref().getLoc(), subspanOp.getType(), subspanOp.binding(), zero,
+        subspanOp.byte_length());
+
+    MLIRContext *context = rewriter.getContext();
+    AffineExpr sym0, sym1;
+    bindSymbols(context, sym0, sym1);
+    auto addMap = AffineMap::get(0, 2, {sym0 + sym1}, context);
+    // We assume that upper layers guarantee the byte offset is perfectly
+    // divisible by the element byte count so the content is well aligned.
+    auto divMap = AffineMap::get(0, 2, {sym0.floorDiv(sym1)}, context);
+
+    // Calculate the offset we need to add to the load/store op, in terms of how
+    // many elements.
+    Optional<int64_t> numBytes = getNumBytes(memrefType.getElementType());
+    if (!numBytes) return failure();
```

Comment:
[optional] return `rewriter.notifyFailure`? I found that this is useful during debugging. Like it will tell you where does it exist.